File: terraform/aws/s3/python/__main__.py
import pulumi
import pulumi_aws as aws

bucket = aws.s3.BucketV2("bucket", bucket="lbriggs-example-website")
cloudfront = aws.cloudfront.Distribution(
    "cloudfront",
    enabled=True,
    wait_for_deployment=False,
    origins=[
        aws.cloudfront.DistributionOriginArgs(
            origin_id="lbriggs-example-website-origin",
            domain_name=bucket.bucket_regional_domain_name,
            custom_origin_config=aws.cloudfront.DistributionOriginCustomOriginConfigArgs(
                http_port=80,
                https_port=443,
                origin_protocol_policy="http-only",
                origin_ssl_protocols=["TLSv1"],
            ),
        )
    ],
    default_cache_behavior=aws.cloudfront.DistributionDefaultCacheBehaviorArgs(
        target_origin_id="lbriggs-example-website-origin",
        allowed_methods=[
            "GET",
            "HEAD",
        ],
        cached_methods=[
            "GET",
            "HEAD",
        ],
        forwarded_values=aws.cloudfront.DistributionDefaultCacheBehaviorForwardedValuesArgs(
            query_string=True,
            cookies=aws.cloudfront.DistributionDefaultCacheBehaviorForwardedValuesCookiesArgs(
                forward="all",
            ),
        ),
        viewer_protocol_policy="redirect-to-https",
        min_ttl=0,
        default_ttl=0,
        max_ttl=0,
    ),
    restrictions=aws.cloudfront.DistributionRestrictionsArgs(
        geo_restriction=aws.cloudfront.DistributionRestrictionsGeoRestrictionArgs(
            restriction_type="none",
        ),
    ),
    viewer_certificate=aws.cloudfront.DistributionViewerCertificateArgs(
        cloudfront_default_certificate=True,
    ),
    price_class="PriceClass_200",
)
briggs = aws.route53.get_zone_output(name="aws.briggs.work.")
root_domain = aws.route53.Record(
    "root_domain",
    zone_id=briggs.zone_id,
    name=briggs.apply(lambda briggs: f"www.{briggs.name}"),
    type="A",
    aliases=[
        aws.route53.RecordAliasArgs(
            name=cloudfront.domain_name,
            zone_id=cloudfront.hosted_zone_id,
            evaluate_target_health=False,
        )
    ],
)
policydoc = aws.iam.get_policy_document_output(
    statements=[
        aws.iam.GetPolicyDocumentStatementArgs(
            sid="AllowCloudFrontServicePrincipal",
            actions=["s3:GetObject"],
            resources=[bucket.arn.apply(lambda arn: f"{arn}/*")],
            conditions=[
                aws.iam.GetPolicyDocumentStatementConditionArgs(
                    test="StringEquals",
                    variable="AWS:SourceArn",
                    values=[cloudfront.arn],
                )
            ],
            principals=[
                aws.iam.GetPolicyDocumentStatementPrincipalArgs(
                    type="Service",
                    identifiers=["cloudfront.amazonaws.com"],
                )
            ],
        )
    ]
)
policy = aws.s3.BucketPolicy("policy", bucket=bucket.id, policy=policydoc.json)
wc = aws.s3.BucketWebsiteConfigurationV2(
    "wc",
    bucket=bucket.bucket,
    index_document=aws.s3.BucketWebsiteConfigurationV2IndexDocumentArgs(
        suffix="index.html",
    ),
    error_document=aws.s3.BucketWebsiteConfigurationV2ErrorDocumentArgs(
        key="index.html",
    ),
    routing_rule_details="""[{
    "Condition": {
        "KeyPrefixEquals": "docs/"
    },
    "Redirect": {
        "ReplaceKeyPrefixWith": ""
    }
}]
""",
)
versioning = aws.s3.BucketVersioningV2(
    "versioning",
    bucket=bucket.id,
    versioning_configuration=aws.s3.BucketVersioningV2VersioningConfigurationArgs(
        status="Enabled",
    ),
)
public_access_block = aws.s3.BucketPublicAccessBlock(
    "public_access_block",
    bucket=bucket.id,
    block_public_acls=True,
    block_public_policy=True,
    ignore_public_acls=True,
    restrict_public_buckets=True,
)
pulumi.export("cloudfrontUrl", cloudfront.domain_name)
pulumi.export("s3Url", wc.website_endpoint)

File: terraform/aws/s3/go/main.go
package main

import (
	"fmt"

	"github.com/pulumi/pulumi-aws/sdk/v6/go/aws/cloudfront"
	"github.com/pulumi/pulumi-aws/sdk/v6/go/aws/iam"
	"github.com/pulumi/pulumi-aws/sdk/v6/go/aws/route53"
	"github.com/pulumi/pulumi-aws/sdk/v6/go/aws/s3"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {
		bucket, err := s3.NewBucketV2(ctx, "bucket", &s3.BucketV2Args{
			Bucket: pulumi.String("lbriggs-example-website"),
		})
		if err != nil {
			return err
		}
		cloudfront, err := cloudfront.NewDistribution(ctx, "cloudfront", &cloudfront.DistributionArgs{
			Enabled:           pulumi.Bool(true),
			WaitForDeployment: pulumi.Bool(false),
			Origins: cloudfront.DistributionOriginArray{
				&cloudfront.DistributionOriginArgs{
					OriginId:   pulumi.String("lbriggs-example-website-origin"),
					DomainName: bucket.BucketRegionalDomainName,
					CustomOriginConfig: &cloudfront.DistributionOriginCustomOriginConfigArgs{
						HttpPort:             pulumi.Int(80),
						HttpsPort:            pulumi.Int(443),
						OriginProtocolPolicy: pulumi.String("http-only"),
						OriginSslProtocols: pulumi.StringArray{
							pulumi.String("TLSv1"),
						},
					},
				},
			},
			DefaultCacheBehavior: &cloudfront.DistributionDefaultCacheBehaviorArgs{
				TargetOriginId: pulumi.String("lbriggs-example-website-origin"),
				AllowedMethods: pulumi.StringArray{
					pulumi.String("GET"),
					pulumi.String("HEAD"),
				},
				CachedMethods: pulumi.StringArray{
					pulumi.String("GET"),
					pulumi.String("HEAD"),
				},
				ForwardedValues: &cloudfront.DistributionDefaultCacheBehaviorForwardedValuesArgs{
					QueryString: pulumi.Bool(true),
					Cookies: &cloudfront.DistributionDefaultCacheBehaviorForwardedValuesCookiesArgs{
						Forward: pulumi.String("all"),
					},
				},
				ViewerProtocolPolicy: pulumi.String("redirect-to-https"),
				MinTtl:               pulumi.Int(0),
				DefaultTtl:           pulumi.Int(0),
				MaxTtl:               pulumi.Int(0),
			},
			Restrictions: &cloudfront.DistributionRestrictionsArgs{
				GeoRestriction: &cloudfront.DistributionRestrictionsGeoRestrictionArgs{
					RestrictionType: pulumi.String("none"),
				},
			},
			ViewerCertificate: &cloudfront.DistributionViewerCertificateArgs{
				CloudfrontDefaultCertificate: pulumi.Bool(true),
			},
			PriceClass: pulumi.String("PriceClass_200"),
		})
		if err != nil {
			return err
		}
		briggs, err := route53.LookupZone(ctx, &route53.LookupZoneArgs{
			Name: pulumi.StringRef("aws.briggs.work."),
		}, nil)
		if err != nil {
			return err
		}
		_, err = route53.NewRecord(ctx, "root_domain", &route53.RecordArgs{
			ZoneId: *pulumi.String(briggs.ZoneId),
			Name:   pulumi.String(fmt.Sprintf("www.%v", briggs.Name)),
			Type:   pulumi.String("A"),
			Aliases: route53.RecordAliasArray{
				&route53.RecordAliasArgs{
					Name:                 cloudfront.DomainName,
					ZoneId:               cloudfront.HostedZoneId,
					EvaluateTargetHealth: pulumi.Bool(false),
				},
			},
		})
		if err != nil {
			return err
		}
		policydoc := iam.GetPolicyDocumentOutput(ctx, iam.GetPolicyDocumentOutputArgs{
			Statements: iam.GetPolicyDocumentStatementArray{
				&iam.GetPolicyDocumentStatementArgs{
					Sid: pulumi.String("AllowCloudFrontServicePrincipal"),
					Actions: pulumi.StringArray{
						pulumi.String("s3:GetObject"),
					},
					Resources: pulumi.StringArray{
						bucket.Arn.ApplyT(func(arn string) (string, error) {
							return fmt.Sprintf("%v/*", arn), nil
						}).(pulumi.StringOutput),
					},
					Conditions: iam.GetPolicyDocumentStatementConditionArray{
						&iam.GetPolicyDocumentStatementConditionArgs{
							Test:     pulumi.String("StringEquals"),
							Variable: pulumi.String("AWS:SourceArn"),
							Values: pulumi.StringArray{
								cloudfront.Arn,
							},
						},
					},
					Principals: iam.GetPolicyDocumentStatementPrincipalArray{
						&iam.GetPolicyDocumentStatementPrincipalArgs{
							Type: pulumi.String("Service"),
							Identifiers: pulumi.StringArray{
								pulumi.String("cloudfront.amazonaws.com"),
							},
						},
					},
				},
			},
		}, nil)
		_, err = s3.NewBucketPolicy(ctx, "policy", &s3.BucketPolicyArgs{
			Bucket: bucket.ID(),
			Policy: policydoc.ApplyT(func(policydoc iam.GetPolicyDocumentResult) (*string, error) {
				return &policydoc.Json, nil
			}).(pulumi.StringPtrOutput),
		})
		if err != nil {
			return err
		}
		wc, err := s3.NewBucketWebsiteConfigurationV2(ctx, "wc", &s3.BucketWebsiteConfigurationV2Args{
			Bucket: bucket.Bucket,
			IndexDocument: &s3.BucketWebsiteConfigurationV2IndexDocumentArgs{
				Suffix: pulumi.String("index.html"),
			},
			ErrorDocument: &s3.BucketWebsiteConfigurationV2ErrorDocumentArgs{
				Key: pulumi.String("index.html"),
			},
			RoutingRuleDetails: pulumi.String(`[{
    "Condition": {
        "KeyPrefixEquals": "docs/"
    },
    "Redirect": {
        "ReplaceKeyPrefixWith": ""
    }
}]
`),
		})
		if err != nil {
			return err
		}
		_, err = s3.NewBucketVersioningV2(ctx, "versioning", &s3.BucketVersioningV2Args{
			Bucket: bucket.ID(),
			VersioningConfiguration: &s3.BucketVersioningV2VersioningConfigurationArgs{
				Status: pulumi.String("Enabled"),
			},
		})
		if err != nil {
			return err
		}
		_, err = s3.NewBucketPublicAccessBlock(ctx, "public_access_block", &s3.BucketPublicAccessBlockArgs{
			Bucket:                bucket.ID(),
			BlockPublicAcls:       pulumi.Bool(true),
			BlockPublicPolicy:     pulumi.Bool(true),
			IgnorePublicAcls:      pulumi.Bool(true),
			RestrictPublicBuckets: pulumi.Bool(true),
		})
		if err != nil {
			return err
		}
		ctx.Export("cloudfrontUrl", cloudfront.DomainName)
		ctx.Export("s3Url", wc.WebsiteEndpoint)
		return nil
	})
}
